cblox_ros/src/tsdf_submap_server.cc

```diff
@@ -0,0 +1,404 @@
+                    num_integrated_frames_per_submap_);
+}
+
+void TsdfSubmapServer::pointcloudCallback(
+    const sensor_msgs::PointCloud2::Ptr& pointcloud_msg_in) {
+  
+  // Pushing this message onto the queue for processing
+  if (pointcloud_msg_in->header.stamp - last_msg_time_ptcloud_ >
+      min_time_between_msgs_) {
+    last_msg_time_ptcloud_ = pointcloud_msg_in->header.stamp;
+    pointcloud_queue_.push(pointcloud_msg_in);
+  }
+
+  // Processing the queue
+  // NOTE(alexmilane): T_G_C - Transformation between Camera frame (C) and
+  //                           global tracking frame (G).
+  Transformation T_G_C;
+  sensor_msgs::PointCloud2::Ptr pointcloud_msg;
+  bool processed_any = false;
+  while (
+      getNextPointcloudFromQueue(&pointcloud_queue_, &pointcloud_msg, &T_G_C)) {
+    constexpr bool is_freespace_pointcloud = false;
+
+    processPointCloudMessageAndInsert(pointcloud_msg, T_G_C,
+                                      is_freespace_pointcloud);
+
+    if (newSubmapRequired()) {
+      createNewSubMap(T_G_C);
+    }
+
+    trajectory_visualizer_ptr_->addPose(T_G_C);
+    visualizeTrajectory();
+
+    processed_any = true;
+  }
+
+  if (!processed_any) {
+    return;
+  }
+
+  if (false) {
+    ROS_INFO_STREAM("Timings: " << std::endl << timing::Timing::Print());
+    // NOTE(alexmillane): This prints the current submap memory. Should update
+    // to print full collection memory.
+    ROS_INFO_STREAM("Active submap memory: " << tsdf_submap_collection_ptr_
+                                                    ->getActiveTsdfMap()
+                                                    .getTsdfLayer()
+                                                    .getMemorySize());
+  }
+}
+
+bool TsdfSubmapServer::getNextPointcloudFromQueue(
+    std::queue<sensor_msgs::PointCloud2::Ptr>* queue,
+    sensor_msgs::PointCloud2::Ptr* pointcloud_msg, Transformation* T_G_C) {
+  const size_t kMaxQueueSize = 10;
+  if (queue->empty()) {
+    return false;
+  }
+  *pointcloud_msg = queue->front();
+  if (transformer_.lookupTransform((*pointcloud_msg)->header.frame_id,
+                                   world_frame_,
+                                   (*pointcloud_msg)->header.stamp, T_G_C)) {
+    queue->pop();
+    return true;
+  } else {
+    if (queue->size() >= kMaxQueueSize) {
+      ROS_ERROR_THROTTLE(60,
+                         "Input pointcloud queue getting too long! Dropping "
+                         "some pointclouds. Either unable to look up transform "
+                         "timestamps or the processing is taking too long.");
+      while (queue->size() >= kMaxQueueSize) {
+        queue->pop();
+      }
+    }
+  }
+  return false;
+}
+
+void TsdfSubmapServer::processPointCloudMessageAndInsert(
+    const sensor_msgs::PointCloud2::Ptr& pointcloud_msg,
+    const Transformation& T_G_C, const bool is_freespace_pointcloud) {
+  // Convert the PCL pointcloud into our awesome format.
```

Comment:
do all of this in another method, in another file. It really clutters things here.